velox/exec/MergingVectorOutput.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/type/Type.h"
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox::exec {
+
+class MergingVectorOutput {
+ public:
+  MergingVectorOutput(
+      velox::memory::MemoryPool* pool,
+      int64_t preferredOutputBatchBytes,
+      int32_t preferredOutputRows,
+      int32_t minOutputBatchRows);
+
+  void addVector(RowVectorPtr input);
+
+  RowVectorPtr getOutput();
+
+  bool needsInput() const {
+    return !noMoreInput_ && outputQueue_.empty();
+  }
+
+  void noMoreInput() {
+    noMoreInput_ = true;
+  }
+
+  bool isFinished() {
+    return noMoreInput_ && bufferRows_ == 0 && outputQueue_.empty();
+  }
+
+ private:
+  // Push the buffer vector to the queue, and reset bufferInputs_,
+  // bufferBytes_ and bufferRows_.
+  void flush();
+
+  // If the input vector is small enough, copy it to the buffer vector.
+  void buffer(RowVectorPtr input);
+
+  velox::memory::MemoryPool* pool_;
+
+  // The preferred output bytes of the bufferInputs_.
+  const int64_t preferredOutputBatchBytes_;
+
+  // The preferred output row count of the bufferInputs_.
+  const int32_t preferredOutputBatchRows_;
+
+  // If the input vector row count is larger than minOutputBatchBytes_, flush
+  // it to the outputQueue_ directly.
+  const int32_t minOutputBatchRows_;
+
+  // The RowVectorPtr queue where bufferInputs_ flush to.
+  std::queue<RowVectorPtr> outputQueue_;
```

Comment:
Please explain why this is FIFO. Why doesn't a vector suffice ?

Reply:
At first I thought it is necessary to ensure the input and output order of vectors, but now it seems not necessary, because only the output of OrderBy needs to ensure the order, but the output of OrderBy will not be merged. So vector can also suffice.

Reply:
It is actually important to preserve the order of the rows. Some operators, like left join preserve the order of inputs rows and some query planners (Presto) take advantage of that.